components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java

```diff
@@ -1239,6 +1276,62 @@ public Response apisApiIdMonetizePost(String apiId, APIMonetizationInfoDTO body,
         return Response.serverError().build();
     }
 
+    /**
+     * Publish API to given external stores.
+     *
+     * @param apiId API Id
+     * @param externalStoreIds  External Store Ids
+     * @param ifMatch   If-match header value
+     * @param messageContext CXF Message Context
+     * @return Response of published external store list
+     */
+    @Override
+    public Response apisApiIdPublishToExternalStoresPost(String apiId, List<String> externalStoreIds, String ifMatch,
+                                                         MessageContext messageContext) {
+        try {
+            String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
+            APIProvider apiProvider = RestApiUtil.getLoggedInUserProvider();
+            APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifierFromUUID(apiId, tenantDomain);
+            API api = apiProvider.getAPI(apiIdentifier);
+            Set<APIStore> inputStores = new HashSet<>();
+            for (String store : externalStoreIds) {
+                if (StringUtils.isNotEmpty(store)) {
+                    APIStore inputStore = APIUtil.getExternalAPIStore(store,
+                            APIUtil.getTenantIdFromTenantDomain(tenantDomain));
+                    if (inputStore == null) {
+                        String errorMessage = "Invalid External Store Id: " + store;
+                        RestApiUtil.handleBadRequest(errorMessage, log);
+                    }
+                    inputStores.add(inputStore);
+                }
+            }
+            Set<String> versions = apiProvider.getAPIVersions(apiIdentifier.getProviderName(),
+                    apiIdentifier.getName());
+            APIVersionStringComparator comparator = new APIVersionStringComparator();
+            boolean apiOlderVersionExist = false;
+            for (String tempVersion : versions) {
+                if (comparator.compare(tempVersion, apiIdentifier.getVersion()) < 0) {
+                    apiOlderVersionExist = true;
+                    break;
+                }
+            }
+            if (apiProvider.updateAPIsInExternalAPIStores(api, inputStores, apiOlderVersionExist)) {
+                Set<APIStore> publishedStores = apiProvider.getPublishedExternalAPIStores(apiIdentifier);
+                ExternalStoreListDTO externalStoreListDTO =
+                        ExternalStoreMappingUtil.fromExternalStoreCollectionToDTO(publishedStores);
+                return Response.ok().entity(externalStoreListDTO).build();
+            }
```

Comment:
Line no 1296 to 1323 seems to be business logic. It would be best we can create a new method in carbon.apimgt.impl layer and call that method from REST API layer instead. Ideally REST API layer shouldn't have any business logics and only some mappings from REST API to impl layer and some REST API specific actions (if necessary). Shall we review other places as well and fix if required?

Reply:
Fixed with https://github.com/wso2/carbon-apimgt/pull/6945/commits/e6d05ebb8d0b632ea87d0651379f4a5495964938#diff-3b19b8a2850e1e1f54616622b325adfbL1034

Moved the same logic to a new method in APIPublisher, **boolean publishToExternalAPIStores(API api, List<String> externalStoreIds)** [https://github.com/wso2/carbon-apimgt/pull/6945/commits/e6d05ebb8d0b632ea87d0651379f4a5495964938#diff-0ed6c08fde442cf700f47efed3dead44R596]